### ISSUE TITLE
关于deepfm的FM()输入的疑问

### ISSUE THREAD

**Issue by DoubleYing** [shenweichen/DeepCTR#155]
**关于deepfm的FM()输入的疑问**
计算linear_logit和dnn_logit时的输入都是features(sparse+dense)，但是fm_logit的输入只有sparse feature，我不知道这里是不是您写错了。还是我没有正确理解deepfm。

**相关代码**
`
linear_logit = get_linear_logit(features, linear_feature_columns, init_std=init_std, seed=seed, prefix='linear',
                                    l2_reg=l2_reg_linear)

    fm_input = concat_fun(sparse_embedding_list, axis=1)   # [B, spare_feat_num, emb_size]
    fm_logit = FM()(fm_input)

    dnn_input = combined_dnn_input(sparse_embedding_list,dense_value_list)  # [B, s+d]
    dnn_out = DNN(dnn_hidden_units, dnn_activation, l2_reg_dnn, dnn_dropout,
                   dnn_use_bn, seed)(dnn_input)
    dnn_logit = tf.keras.layers.Dense(
        1, use_bias=False, activation=None)(dnn_out)
`

**Comment by shenweichen** [comment]
densefeat 需要进行离散化作为 sparsefeat 才可以参与到fm组件的特征交互中。具体可以参考我之前的一个知乎回答 https://www.zhihu.com/question/352399723/answer/869939360

**Comment by DoubleYing** [comment]
感谢作者的回复，看了您在知乎上的回答的4种方式来处理density feat，但是这几种方式最后能够达到的效果是一样的嘛？它们的适用场景一样吗？还是说需要根据实际的数据来做各种尝试以后才能确定哪个才是最佳的。。

**Comment by lisency** [comment]
最好的方式还是将所有连续值的特征做离散化，这个版本的DeepFM没有做离散化，导致MF部分无法捕捉到连续特征的order-2参数。

另外，这个实现当中LR部分的输入是对所有sparse特征的embedding做sum，这与论文里有些出入，论文里对LR学习的是单独的参数W。比如对例子程序输入是26 * 4，实际上应该是26 * 1。这导致这个版本无法学习到memorization信息。

![image](https://user-images.githubusercontent.com/8286834/69960130-b0b70d00-1543-11ea-83fa-817835d2aeb6.png)

**Comment by lonewar** [comment]
> 最好的方式还是将所有连续值的特征做离散化，这个版本的DeepFM没有做离散化，导致MF部分无法捕捉到连续特征的order-2参数。
> 
> 另外，这个实现当中LR部分的输入是对所有sparse特征的embedding做sum，这与论文里有些出入，论文里对LR学习的是单独的参数W。比如对例子程序输入是26 * 4，实际上应该是26 * 1。这导致这个版本无法学习到memorization信息。
> 
> ![image](https://user-images.githubusercontent.com/8286834/69960130-b0b70d00-1543-11ea-83fa-817835d2aeb6.png)

第二个没有问题吧？

**Comment by DeveloperMrYang** [comment]
> 论文里对LR学习的是单独的参数W

+1，也许这么做是为了泛化性？

**Comment by DeveloperMrYang** [comment]
> 最好的方式还是将所有连续值的特征做离散化，这个版本的DeepFM没有做离散化，导致MF部分无法捕捉到连续特征的order-2参数。
> 
> 另外，这个实现当中LR部分的输入是对所有sparse特征的embedding做sum，这与论文里有些出入，论文里对LR学习的是单独的参数W。比如对例子程序输入是26 * 4，实际上应该是26 * 1。这导致这个版本无法学习到memorization信息。
> 
> ![image](https://user-images.githubusercontent.com/8286834/69960130-b0b70d00-1543-11ea-83fa-817835d2aeb6.png)

+1，也许这么做是为了泛化性？

**Comment by DeveloperMrYang** [comment]
> 最好的方式还是将所有连续值的特征做离散化，这个版本的DeepFM没有做离散化，导致MF部分无法捕捉到连续特征的order-2参数。
> 
> 另外，这个实现当中LR部分的输入是对所有sparse特征的embedding做sum，这与论文里有些出入，论文里对LR学习的是单独的参数W。比如对例子程序输入是26 * 4，实际上应该是26 * 1。这导致这个版本无法学习到memorization信息。
> 
> ![image](https://user-images.githubusercontent.com/8286834/69960130-b0b70d00-1543-11ea-83fa-817835d2aeb6.png)

26×1的话，离散特征应该用

> 最好的方式还是将所有连续值的特征做离散化，这个版本的DeepFM没有做离散化，导致MF部分无法捕捉到连续特征的order-2参数。
> 
> 另外，这个实现当中LR部分的输入是对所有sparse特征的embedding做sum，这与论文里有些出入，论文里对LR学习的是单独的参数W。比如对例子程序输入是26 * 4，实际上应该是26 * 1。这导致这个版本无法学习到memorization信息。
> 
> ![image](https://user-images.githubusercontent.com/8286834/69960130-b0b70d00-1543-11ea-83fa-817835d2aeb6.png)

您的意思是指，在将离散特征LabelEnconder作为Input后，直接加个输出单元为1的全连接层？

**Comment by DeveloperMrYang** [comment]
> 最好的方式还是将所有连续值的特征做离散化，这个版本的DeepFM没有做离散化，导致MF部分无法捕捉到连续特征的order-2参数。
> 
> 另外，这个实现当中LR部分的输入是对所有sparse特征的embedding做sum，这与论文里有些出入，论文里对LR学习的是单独的参数W。比如对例子程序输入是26 * 4，实际上应该是26 * 1。这导致这个版本无法学习到memorization信息。
> 
> ![image](https://user-images.githubusercontent.com/8286834/69960130-b0b70d00-1543-11ea-83fa-817835d2aeb6.png)

您的意思是指，在将离散特征LabelEnconder作为Input后，直接加个输出单元为1的全连接层？

**Comment by zanshuxun** [comment]
> 最好的方式还是将所有连续值的特征做离散化，这个版本的DeepFM没有做离散化，导致MF部分无法捕捉到连续特征的order-2参数。
> 
> 另外，这个实现当中LR部分的输入是对所有sparse特征的embedding做sum，这与论文里有些出入，论文里对LR学习的是单独的参数W。比如对例子程序输入是26 * 4，实际上应该是26 * 1。这导致这个版本无法学习到memorization信息。
> 
> ![image](https://user-images.githubusercontent.com/8286834/69960130-b0b70d00-1543-11ea-83fa-817835d2aeb6.png)

linear部分的输入哪里有26*4？

**Comment by guotong1988** [comment]
学习了
